toolchain/parse/node_ids.h

```diff
@@ -0,0 +1,96 @@
+#include "toolchain/parse/node_kind.h"
+
+namespace Carbon::Parse {
+
+// A lightweight handle representing a node in the tree.
+//
+// Objects of this type are small and cheap to copy and store. They don't
+// contain any of the information about the node, and serve as a handle that
+// can be used with the underlying tree to query for detailed information.
+struct NodeId : public IdBase {
+  // An explicitly invalid instance.
+  static const NodeId Invalid;
+
+  using IdBase::IdBase;
+};
+
+constexpr NodeId NodeId::Invalid = NodeId(NodeId::InvalidIndex);
+
+// For looking up the type associated with a given id type.
+template <typename T>
+struct NodeForId;
+
+// `<KindName>Id` is a typed version of `NodeId` that references a node of kind
+// `<KindName>`:
+template <const NodeKind&>
+struct NodeIdForKind : public NodeId {
+  static const NodeIdForKind Invalid;
+
+  explicit NodeIdForKind(NodeId node_id) : NodeId(node_id) {}
+};
+template <const NodeKind& Kind>
+constexpr NodeIdForKind<Kind> NodeIdForKind<Kind>::Invalid =
+    NodeIdForKind(NodeId::Invalid.index);
+
+#define CARBON_PARSE_NODE_KIND(KindName) \
+  using KindName##Id = NodeIdForKind<NodeKind::KindName>;
+#include "toolchain/parse/node_kind.def"
+
+// NodeId that matches any NodeKind whose `category()` overlaps with `Category`.
+template <NodeCategory Category>
+struct NodeIdInCategory : public NodeId {
+  // An explicitly invalid instance.
+  static const NodeIdInCategory<Category> Invalid;
+
+  explicit NodeIdInCategory(NodeId node_id) : NodeId(node_id) {}
+};
+
+template <NodeCategory Category>
+constexpr NodeIdInCategory<Category> NodeIdInCategory<Category>::Invalid =
+    NodeIdInCategory<Category>(NodeId::InvalidIndex);
+
+// Aliases for `NodeIdInCategory` to describe particular categories of nodes.
+using AnyDecl = NodeIdInCategory<NodeCategory::Decl>;
+using AnyExpr = NodeIdInCategory<NodeCategory::Expr>;
+using AnyModifier = NodeIdInCategory<NodeCategory::Modifier>;
+using AnyNameComponent = NodeIdInCategory<NodeCategory::NameComponent>;
+using AnyPattern = NodeIdInCategory<NodeCategory::Pattern>;
+using AnyStatement = NodeIdInCategory<NodeCategory::Statement>;
+
+// NodeId with kind that matches either T::Kind or U::Kind.
+template <typename T, typename U>
+struct NodeIdOneOf : public NodeId {
+  // An explicitly invalid instance.
+  static const NodeIdOneOf<T, U> Invalid;
+
+  explicit NodeIdOneOf(NodeId node_id) : NodeId(node_id) {}
+};
+
+template <typename T, typename U>
+constexpr NodeIdOneOf<T, U> NodeIdOneOf<T, U>::Invalid =
+    NodeIdOneOf<T, U>(NodeId::InvalidIndex);
+
+// NodeId with kind that is anything but T::Kind.
+template <typename T>
+struct NodeIdNot : public NodeId {
+  // An explicitly invalid instance.
+  static const NodeIdNot<T> Invalid;
+
+  explicit NodeIdNot(NodeId node_id) : NodeId(node_id) {}
+};
+
+template <typename T>
+constexpr NodeIdNot<T> NodeIdNot<T>::Invalid =
+    NodeIdNot<T>(NodeId::InvalidIndex);
```

Comment:
Would it make sense to take `const NodeKind&` here instead of `typename`? All extract wants is the kind, and passing the kind in will mean that these can be used in `typed_nodes.h` before the corresponding typed node is defined. I think this is the only remaining reason why `typed_nodes.h` needs to be in a somewhat-topological order.

Reply:
OK, let's leave as-is for now.